src/backend/distributed/master/master_node_protocol.c

```diff
@@ -801,6 +806,66 @@ GetTableForeignConstraintCommands(Oid relationId)
 
 
 /*
+ * GenerateForeignConstraintCommandWithNotValid takes a ALTER TABLE ... ADD FOREIGN KEY
+ * command and generates a new command which is exactly the same, other than the
+ * prepended NOT VALID part, which allows creation of the constraint with the validation
+ * step skipped.
+ *
+ * The function requires that the given command is an ALTER TABLE query with only one
+ * ADD FOREIGN KEY command.
+ */
+char *
+GenerateForeignConstraintCommandWithNotValid(char *constraintCommand)
+{
+	Node *parsetree = ParseTreeNode(constraintCommand);
+	AlterTableStmt *alterTableStatemet = NULL;
+	List *commandList = NIL;
+	AlterTableCmd *alterTableCommand = NULL;
+	Constraint *constraint = NULL;
+	StringInfo commandWithNotValid = makeStringInfo();
+
+	Assert(IsA(parsetree, AlterTableStmt));
+
+	alterTableStatemet = (AlterTableStmt *) parsetree;
+	commandList = alterTableStatemet->cmds;
+
+	if (list_length(commandList) > 1)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("ALTER TABLE .. FOREIGN KEY statements with "
+							   "more than one command are not supported.")));
+	}
+
+	alterTableCommand = (AlterTableCmd *) lfirst(commandList->head);
+	constraint = (Constraint *) alterTableCommand->def;
+
+	Assert(alterTableCommand->subtype == AT_AddConstraint);
+	Assert(constraint->contype == CONSTR_FOREIGN);
+
+	appendStringInfo(commandWithNotValid, "%s", constraintCommand);
+
+	/* add the NOT VALID clause if the original command does not containts it */
+	if (constraint->skip_validation != true)
+	{
+		/* strip any trailing spaces and semicolons */
+		int length = commandWithNotValid->len;
+		while (isspace((unsigned char) commandWithNotValid->data[length - 1]) ||
+			   commandWithNotValid->data[length - 1] == ';')
+		{
+			commandWithNotValid->data[length - 1] = '\0';
+			length--;
+		}
+
+		commandWithNotValid->len = length;
+
+		appendStringInfo(commandWithNotValid, " NOT VALID");
```

Comment:
I quite like the idea of just modifying the foreign key to be `NOT VALID`, though we currently alter the workers before altering the master so this might be a bigger change. Considering all the options, appending " NOT VALID" might be the best way to keep us moving forward.